bus/accounts.go

```diff
@@ -0,0 +1,125 @@
+package bus
+
+import (
+	"math/big"
+	"sync"
+
+	"go.sia.tech/core/types"
+	"go.sia.tech/renterd/ephemeralaccounts"
+	rhpv3 "go.sia.tech/renterd/rhp/v3"
+)
+
+type accounts struct {
+	mu      sync.Mutex
+	byID    map[rhpv3.Account]*account
+	byOwner map[string][]*account
+}
+
+type account struct {
+	mu sync.Mutex
+	ephemeralaccounts.Account
+	Owner string
+}
+
+func newAccounts(accs []ephemeralaccounts.Account) *accounts {
+	a := &accounts{
+		byID:    make(map[rhpv3.Account]*account),
+		byOwner: make(map[string][]*account),
+	}
+	for _, acc := range accs {
+		account := &account{
+			Account: ephemeralaccounts.Account{
+				ID:      acc.ID,
+				Host:    acc.Host,
+				Balance: acc.Balance,
+			},
+			Owner: acc.Owner,
+		}
+		a.byID[account.ID] = account
+		a.byOwner[acc.Owner] = append(a.byOwner[acc.Owner], account)
+	}
+	return a
+}
+
+// AddAmount apples the provided amount to an account through addition. So the
+// input can be both a positive or negative number depending on whether a
+// withdrawal or deposit is recorded. If the account doesn't exist, it is
+// created.
+func (a *accounts) AddAmount(id rhpv3.Account, owner string, hk types.PublicKey, amt *big.Int) {
+	acc := a.account(id, owner, hk)
+
+	// Update balance.
+	acc.mu.Lock()
+	acc.Balance.Add(acc.Balance, amt)
+	acc.mu.Unlock()
+}
+
+// UpdateBalance sets the balance of a given account to the provided amount. If
+// the account doesn't exist, it is created.
+func (a *accounts) UpdateBalance(id rhpv3.Account, owner string, hk types.PublicKey, balance *big.Int) {
+	acc := a.account(id, owner, hk)
+
+	// Update balance.
+	acc.mu.Lock()
+	acc.Balance = balance
+	acc.mu.Unlock()
+}
+
+// Accounts returns all accounts for a given owner. Usually called when workers
+// request their accounts on startup.
+func (a *accounts) Accounts(owner string) []ephemeralaccounts.Account {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	accounts := make([]ephemeralaccounts.Account, len(a.byOwner[owner]))
+	for i, acc := range a.byOwner[owner] {
+		acc.mu.Lock()
```

Comment:
double lock, do we care?

Reply:
As long as the locking order is correct this should be fine. Especially since these methods are hardly ever called.